dex/networks/zec/tx.go

```diff
@@ -0,0 +1,417 @@
+		}
+	}
+
+	if err := wire.WriteVarInt(w, pver, 0); err != nil {
+		return nil, err
+	}
+
+	return w.Bytes(), nil
+}
+
+// see https://zips.z.cash/protocol/protocol.pdf section 7.1
+func DeserializeZecTx(b []byte) (*ZecTx, error) {
+	tx := &ZecTx{MsgTx: new(wire.MsgTx)}
+	r := bytes.NewReader(b)
+	if err := tx.ZecDecode(r); err != nil {
+		return nil, err
+	}
+	return tx, nil
+}
+
+// ZecDecode reads the serialized transaction from the reader and populates the
+// *ZecTx's fields.
+func (tx *ZecTx) ZecDecode(r io.Reader) error {
+	ver, err := readUint32(r)
+	if err != nil {
+		return err
+	}
+
+	// overWintered := (ver & (1 << 31)) > 0
+	ver &= overwinterMask // Clear the overwinter bit
+	tx.Version = int32(ver)
+
+	if ver > 4 {
+		return fmt.Errorf("unsupported tx version %d > 4", ver)
+	}
+
+	if ver > 3 {
+		_, err := readUint32(r)
+		if err != nil {
+			return err
+		}
+	}
+
+	txInCount, err := wire.ReadVarInt(r, pver)
+	if err != nil {
+		return err
+	}
+
+	tx.TxIn = make([]*wire.TxIn, 0, txInCount)
+	for i := 0; i < int(txInCount); i++ {
+		ti := new(wire.TxIn)
+		if err := readTxIn(r, pver, int32(ver), ti); err != nil {
+			return err
+		}
+		tx.TxIn = append(tx.TxIn, ti)
+	}
+
+	txOutCount, err := wire.ReadVarInt(r, pver)
+	if err != nil {
+		return err
+	}
+
+	tx.TxOut = make([]*wire.TxOut, 0, txOutCount)
+	for i := 0; i < int(txOutCount); i++ {
+		to := new(wire.TxOut)
+		if err := readTxOut(r, pver, int32(ver), to); err != nil {
+			return err
+		}
+		tx.TxOut = append(tx.TxOut, to)
+	}
+
+	tx.LockTime, err = readUint32(r)
+	if err != nil {
+		return err
+	}
+
+	if ver > 3 {
+		tx.ExpiryHeight, err = readUint32(r)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Empty the rest of the buffer.
+	if ver < 2 {
+		return nil
+	}
+
+	var bindingSigRequired bool
+	if ver >= 4 {
+		// valueBalanceSpending
+		if _, err := readUint64(r); err != nil {
+			return fmt.Errorf("error reading valueBalanceSpending: %w", err)
+		}
+
+		if nSpendsSapling, err := wire.ReadVarInt(r, pver); err != nil {
+			return fmt.Errorf("error reading nSpendsSapling: %w", err)
+		} else if nSpendsSapling > 0 {
+			// vSpendsSapling - discard
+			bindingSigRequired = true
+			if _, err := io.ReadFull(r, make([]byte, nSpendsSapling*384)); err != nil {
+				return fmt.Errorf("error reading vSpendsSapling: %w", err)
+			}
+		}
+
+		if nOutputsSapling, err := wire.ReadVarInt(r, pver); err != nil {
+			return fmt.Errorf("error reading nOutputsSapling: %w", err)
+		} else if nOutputsSapling > 0 {
+			// vOutputsSapling - discard
+			bindingSigRequired = true
+			if _, err := io.ReadFull(r, make([]byte, nOutputsSapling*948)); err != nil {
+				return fmt.Errorf("error reading vOutputsSapling: %w", err)
+			}
+		}
+
+	}
+
+	if nJoinSplit, err := wire.ReadVarInt(r, pver); err != nil {
+		return fmt.Errorf("error reading nJoinSplit: %w", err)
+	} else if nJoinSplit > 0 {
+		// vJoinSplit - discard
+		sz := 1802 * nJoinSplit
+		if ver == 4 {
+			sz = 1698 * nJoinSplit
```

Comment:
Curious about the meaning of these magics.

Also, this function does a lot of discarding of fairly large slices of data, so to simply seek forward without allocating a junk slice, I'd consider using `buf := bytes.NewBuffer(blk)`, and then `buf.Next(N)` to throw out `N` bytes.  But you'd have to check the length of the returned slice.  There's also a `bufio.Reader` with a `Discard` method that returns an error, but I've not used that one and it might be inefficient for this purpose.

Another thought on that is that a `*bytes.Reader` (the concrete type, not the `io.Reader` interface) such as the one created by `DeserializeZecTx` is also an `io.ReedSeeker` that would just give us the `Seek` method, but that makes the `ZecDecode` input more specialized (e.g. the `io.Reader` returned by `hex.NewDecoder` is not also a `Seeker`).